vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+     * FIXME: need a double-check for the definition
+     */
+    bool edge_is_configurable(const RREdgeId& edge) const;
+    bool edge_is_non_configurable(const RREdgeId& edge) const;
+
+    /* Switch Info */
+    size_t switch_index(const RRSwitchId& switch_id) const; /* TODO: deprecate this accessor as outside functions should use RRSwitchId */
+    /* Get the switch info of a switch used in this RRGraph */
+    const t_rr_switch_inf& get_switch(const RRSwitchId& switch_id) const;
+
+    /* Segment Info */
+    size_t segment_index(const RRSegmentId& segment_id) const; /* TODO: deprecate this accessor as outside functions should use RRSegmentId */
+    /* Get the segment info of a routing segment used in this RRGraph */
+    const t_segment_inf& get_segment(const RRSegmentId& segment_id) const;
+
+    /* Utilities */
+    /* Find the edges connecting two nodes */
+    std::vector<RREdgeId> find_edges(const RRNodeId& src_node, const RRNodeId& sink_node) const;
+    /* Find a node with given features from internal fast look-up */
+    RRNodeId find_node(const short& x, const short& y, const t_rr_type& type, const int& ptc, const e_side& side = NUM_SIDES) const;
+    /* Find the number of routing tracks in a routing channel with a given coordinate */
+    short chan_num_tracks(const short& x, const short& y, const t_rr_type& type) const;
+
+    /* Check if the graph contains invalid nodes/edges etc. */
+    bool is_dirty() const;
+
+  public:                                        /* Echos */
+    void print_node(const RRNodeId& node) const; /* Print the detailed information of a node */
+
+  public: /* Public Checkers */
+    /* Full set checking using listed checking functions*/
+    bool check() const;
+
+  public: /* Public Validators */
+    /* Validate is the node id does exist in the RRGraph */
+    bool valid_node_id(const RRNodeId& node) const;
+    /* Validate is the edge id does exist in the RRGraph */
+    bool valid_edge_id(const RREdgeId& edge) const;
+
+  public: /* Mutators */
+    /* Reserve the lists of nodes, edges, switches etc. to be memory efficient. 
+     * This function is mainly used to reserve memory space inside RRGraph,
+     * when adding a large number of nodes/edge/switches/segments,
+     * in order to avoid memory fragements
+     * For example: 
+     *    reserve_nodes(1000);
+     *    for (size_t i = 0; i < 1000; ++i) {
+     *      create_node();
+     *    }
+     */
+    void reserve_nodes(const int& num_nodes);
+    void reserve_edges(const int& num_edges);
+    void reserve_switches(const int& num_switches);
+    void reserve_segments(const int& num_segments);
+
+    /* Add new elements (node, edge, switch, etc.) to RRGraph */
+    /* Add a node to the RRGraph with a deposited type 
+     * Detailed node-level information should be added using the set_node_* functions
+     * For example: 
+     *   RRNodeId node = create_node();
```

Comment:
Suggest adding an actual rr_graph object to this example so the code shows how it is referenced.